src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs

```diff
@@ -567,12 +569,34 @@ impl Persist for VirtSandbox {
 
     /// Save a state of Sandbox
     async fn save(&self) -> Result<Self::State> {
+        let hypervisor_state = self.hypervisor.save_state().await?;
         let sandbox_state = crate::sandbox_persist::SandboxState {
             sandbox_type: VIRTCONTAINER.to_string(),
             resource: Some(self.resource_manager.save().await?),
-            hypervisor: Some(self.hypervisor.save_state().await?),
+            hypervisor: match hypervisor_state.hypervisor_type.as_str() {
+                // TODO support other hypervisors
+                HYPERVISOR_DRAGONBALL => Ok(Some(hypervisor_state)),
+                HYPERVISOR_FIRECRACKER => Ok(Some(hypervisor_state)),
+                _ => Err(anyhow!(
+                    "Unsupported hypervisor {}",
+                    hypervisor_state.hypervisor_type
+                )),
+            }?,
+        };
+        // FIXME: properly handle jailed case
+        // eg: Determine if we are running jailed:
+        // let h = sandbox_state.hypervisor.clone().unwrap_or_default();
+        // Figure out the jailed path:
+        // jailed_path = h.<>
+        // and somehow store the sandbox state into the jail:
+        // persist::to_disk(&sandbox_state, &self.sid, jailed_path)?;
+        // Issue is, how to handle restore.
```

Comment:
Do I get the issue right that it's the chicken-egg problem where in order to restore a hypervisor you need to get the corresponding jailed path, for which you need the very same hypervisor you're trying to restore?

Reply:
again, correct -- and we don't like the way we've implemented that (at all...) so we're open to suggestions about what to do here ...

Reply:
Honestly I don't know if there can even be any other way apart from either 1) not saving the state to the jailed path, or 2) store the jailed path also somewhere outside of the state (which would require building something of a registry of sandbox state paths...).